numpy/_core/src/multiarray/stringdtype/static_string.c

```diff
@@ -0,0 +1,745 @@
+                       npy_string_allocator *allocator)
+{
+    if (size > NPY_MAX_STRING_SIZE) {
+        return -1;
+    }
+
+    _npy_static_string_u *out_u = (_npy_static_string_u *)out;
+
+    unsigned char flags =
+            out_u->direct_buffer.size_and_flags & ~NPY_SHORT_STRING_SIZE_MASK;
+
+    if (size == 0) {
+        memcpy(out_u, &empty_string_u, sizeof(_npy_static_string_u));
+        out_u->direct_buffer.size_and_flags |= flags;
+        return 0;
+    }
+
+    if (size > NPY_SHORT_STRING_MAX_SIZE) {
+        int on_heap = 0;
+        char *buf = heap_or_arena_allocate(allocator, out_u, size, &on_heap);
+
+        if (buf == NULL) {
+            return -1;
+        }
+
+        if (on_heap) {
+            out_u->vstring.offset = (size_t)buf;
+        }
+        else {
+            npy_string_arena *arena = &allocator->arena;
+            if (arena == NULL) {
+                return -1;
+            }
+            out_u->vstring.offset = (size_t)buf - (size_t)arena->buffer;
+        }
+        set_vstring_size(out_u, size);
+    }
+    else {
+        // Size can be no larger than 7 or 15, depending on CPU architecture.
+        // In either case, the size data is in at most the least significant 4
+        // bits of the byte so it's safe to | with one of 0x10, 0x20, 0x40, or
+        // 0x80.
+        out_u->direct_buffer.size_and_flags = NPY_STRING_SHORT | flags | size;
+    }
+
+    return 0;
+}
+
+NPY_NO_EXPORT int
+NpyString_free(npy_packed_static_string *str, npy_string_allocator *allocator)
+{
+    _npy_static_string_u *str_u = (_npy_static_string_u *)str;
+    if (is_not_a_vstring(str)) {
+        // zero out, keeping flags
+        unsigned char *flags = &str_u->direct_buffer.size_and_flags;
+        unsigned char current_flags = *flags & ~NPY_SHORT_STRING_SIZE_MASK;
+        memcpy(str_u, &empty_string_u, sizeof(_npy_static_string_u));
+        *flags |= current_flags;
+    }
+    else {
+        if (VSTRING_SIZE(str_u) == 0) {
+            // empty string is a vstring but nothing to deallocate
+            return 0;
+        }
+        if (heap_or_arena_deallocate(allocator, str_u) < 0) {
+            return -1;
+        }
+    }
+    return 0;
+}
+
+/*NUMPY_API
+ * Pack the null string into a npy_packed_static_string
+ *
+ * Pack the null string into *packed_string*. Returns 0 on success and -1
+ * on failure.
+*/
+NPY_NO_EXPORT int
+NpyString_pack_null(npy_string_allocator *allocator,
+                    npy_packed_static_string *packed_string)
+{
+    _npy_static_string_u *str_u = (_npy_static_string_u *)packed_string;
+    unsigned char *flags = &str_u->direct_buffer.size_and_flags;
+    unsigned char current_flags = *flags & ~NPY_SHORT_STRING_SIZE_MASK;
+    if (NpyString_free(packed_string, allocator) < 0) {
+        return -1;
+    }
+    memcpy(str_u, &empty_string_u, sizeof(_npy_static_string_u));
+    *flags = current_flags | NPY_STRING_MISSING;
```

Comment:
They need to be preserved because we allow mutation, so we need metadata about the original allocation associated with this string in order to determine if there is space in the arena allocation for a new string after a user mutates this one to a non-NULL value. Added a comment.